### ISSUE TITLE
Support declaring a dependency version as strict

### ISSUE THREAD

**Issue by melix** [gradle/gradle#3168]
### Context

Often, a dependency cannot easily be upgraded. For example, if a component depends on MyLibrary version 4.0.0, it's unlikely that it would work correctly on MyLibrary version 5.0.0, because MyLibrary follows semantic versioning and that 5 is binary incompatible with 4. However, traditional dependency resolution supports upgrading dependency versions. This means that if a transitive dependency pulls in a more recent version, it will result in the dependency version `4` being upgraded to `5`. Another common use case is "this library MUST be used with this other". For example, the `api` and `implementation` of a dependency are bound, and must come with the very same version number.

### Proposed solution

What we propose here is a way to specify that a dependency version is strict. This means that:

- if no other dependency on the same module is found transitively, then the version is selected
- if another dependency on the same module is found, and that the selector for this version is compatible with the strict one, the strict version is selected
- in all other cases, fail

This means, in particular, that:

- if a dependency on version `15` is strict, and that a transitive dependency brings version `17`, the build should fail
- if a dependency on version `[15,20]` is strict, and that a transitive dependency brings version `17`, then version `17` is selected (because it's compatible)
- if a first-level dependency says `17`, but a transitive dependency says strictly `15`, then the build fails
- if a first-level dependency says `17`, but a transitive dependency says `15` (not strict), then the `17` version is used

We will _not_ downgrade dependencies (in particular for the latter case), because dependency management as it exists now is all about assuming backwards compatibility. In other words, we can assume that if you upgrade from `15` to `17`, things are going to work fine, but not the other way around. This story is precisely about this assumption being false, and providing a way to fix it.

This differs from the existing `failOnConflict()` flag in the sense that this is _specific to a dependency_, while `failOnConflict()` would fail _whatever version breaks_.

### Non-goals

- This story doesn't explain _how_ to solve the conflict. It's likely that if you have such a conflict, you have to think about how to solve it because you have 2 dependencies that disagree on something that is supposed to be strict (at least for one of the members). Typical resolution can always happen using the `resolutionStrategy.eachDependency` block.
- The `strict` constraint will not be published (yet). This belongs to a subsequent story. It means that it will work within a multi-project Gradle build for now, but not as soon as a dependency is published.

### Implementation options

The `force` flag that exists already on dependencies also influences how dependencies are resolved. However, it only applies on first-level dependencies. This behavior is inherited from Ivy's `force` flag, and exposed to the Gradle DSL. However, instead of adding another `strict` flag that would conflict with the `force` flag, we recommend to:

- deprecate the `force` flag
- introduce a new `ForceMode` enum that would have 3 possible values:
** `soft`: would allow a dependency version to be upgraded, as it is typically using the latest version conflict resolution strategy. This is the default.
** `force`: would behave like `force=true` today, by forcing the version of first-level dependencies, unless a transitive dependency is `strict`
** `strict`: would implement the strict strategy described in this ticket.

In terms of DSL, this "flag" could live where the `force` flag is seen today. So instead of:

```
dependencies {
   implementation('org:foo:1.0') {
     force = true // now deprecated
   }
}
```

we would write:

```
dependencies {
   implementation('org:foo:1.0') {
     forceMode = ForceMode.force // direct-dependency forcing strategy
   }
}
```

or

```
dependencies {
   implementation('org:foo:1.0') {
     forceMode = ForceMode.strict // strict dependency as described in this ticket
   }
}
```

Any better name than `forceMode` is very welcome.

### Related work

This _could_ serve as the basis for both "recommendations" and "platforms" implementation. One could see a Maven BOM (bill of materials) as a set of recommended versions. That is to say, _if_ a dependency onto a component is found, we should use the version recommended in the BOM. However if a transitive dependency disagrees, it may be upgraded. But, one could also import a BOM as a "platform" requirement, in which case we _could_ use strict dependencies here: if a dependency onto a module is found, then we _must_ use the version found in the BOM, or fail.

_This is WIP spec_

**Comment by adammurdoch** [comment]
It would be good to start with the use cases before designing the implementation. That is, start with the concrete reasons _why_ you might want stricter matching.

**Comment by oehme** [comment]
The most common use case is: As the author of a library you already know that it won't work with anything higher than the given version. I.e. you've already tried it and it failed. Thus you want Gradle to fail if conflict resolution would upgrade it.

Another use case would be that you want to model a "platform", i.e. you want to deploy to some specific application server and this server provides a couple of libraries in fixed versions. Any attempt to upgrade these versions should result in a build failure, as they wouldn't match the deployment target.

A related concept would be for a publisher to be able to say "I break APIs on major releases, so resolution should fail if a dependency on me is upgraded beyond a major version boundary". Even with that capability, strict dependencies can still be useful to solve the case where a consumer needs to narrow the constraint down even further, because the producer didn't think of some side-effect that the consumer depended on.

**Comment by bigdaz** [comment]
I'd like to consider other ways to model this than as an extension to `force`, but something more like a true 'compatibility constraint'.
So instead of saying "I strictly depend on v1.1" you'd say 2 things: "I prefer v1.1" and "I am only compatible with v1.1". This might more easily be adapted to other use cases like semantic versioning, where you would say "I prefer 1.0.0, and I am only compatible with versions in the range [1.1.0, 2.0.0)".

**Comment by melix** [comment]
Yes, this is what I thought when working on publishing steps. It clearly looks like a compatibility constraint. Do you have an idea what it would look like in practice?

**Comment by oehme** [comment]
Another way to model this would be "I'm compatible with [1.0, 3.0)" + "I'm incompatible with [3.0, )". I'd still consider it valuable to have a concise way of saying both these things at once though.

**Comment by oehme** [comment]
@bigdaz I'd consider semantic versioning as something that belongs on the producer, not the consumer. I.e. "foo:1.1" would have metadata saying "You can safely replace me with [1.1, 2.0)".

**Comment by oehme** [comment]
@melix as far as concrete syntax would go, here's one idea:

```
constraints {
  incompatibleWith("com.foo:bar:3.0")
  recommend("com.foo:bar:2.1")
}
```

A `Dependency` might just be a special kind of constraint. I'm not even sure I like the word constraint yet, it's just a placeholder I invented until we come up with something better :)

**Comment by melix** [comment]
Ok if we step back a bit, it's a dependency declaration. A typical dependency says:

- group: org
- name: bar
- version: 1.0

However, we want to _soften_ or _harden_ constraints on the dependency version. One way to do it today is to say:

- version: [1.0, 2.0)

However, it doesn't prevent the dependency from being upgraded, because "version" implies "preferred version". It will not let us fail depending on the conflicts. So we need a richer model, which may imply that "version" alone is not enough (it's the _requested_ version)

**Comment by melix** [comment]
I'm not sure I like:

> incompatibleWith("com.foo:bar:3.0")

does it even make sense? This kind of information tends to be dynamic. At a fixed point in time you only know which versions you are compatible with.

**Comment by oehme** [comment]
I think explicit incompatibility makes sense. It would allow you to attach a proper reason to that constraint for instance. E.g. "This version has a bug that breaks us and which was fixed in the next patch release". Or "I'm incompatible with this component in general, please fail if it is in the graph".

**Comment by oehme** [comment]
So here are three ways one could express this kind of stricter matching. Each of them has something going for it:

```
dependencies {
  implementation ("guava:[15, 17]") {
    //this is the current proposal, it's very simple to understand
    strict = true
  }
}
```

or 

```
dependencies {
  implementation "guava:[15, 17]")
}
constraints {
  //We want incompatibility constraints for other reasons too, so this might supersede "strict"
  incompatibleWith "guava:[17, )" 
}
```

or even

```
dependencies {
  implementation ("guava:17") { //treat the current "version" field as a preference
    //match strictly against this new, Gradle-only information
    compatibleVersions = "[15, 17]"
  }
}
```

**Comment by melix** [comment]
Doc for further discussion: https://docs.google.com/document/d/1iwXPktvQbVFPsEyyiLzEyS6TinmwFExZTLS0rkk_DY4/edit#

**Comment by adammurdoch** [comment]
We should try really hard to not use boolean flags and a string here, as we want a richer language than just "strict" (whatever that is). Instead, we want to model these constraints as strongly typed component selectors (or component constraints if you like).

Currently, a dependency declaration has-a single version selector. We have the following types:

`version='x'` means:

- prefer x
- accept anything >x, must not use anything <x

`version='[x,y]' means:

- prefer the latest version >=x and < y
- accept anything >=x, must not use anything < x

`version='latest.x'` means:

- prefer the latest version with status x.
- accept anything > latest version with status x, must not use anything earlier.

`x+` does string prefix matching which doesn't mean anything useful, but that's another issue.

For the use cases listed in the description, what we want is a way to express the following constraints:

"I require x", which means:

- prefer x
- accept x, must not use anything != x

"I require anything from x up to y" which means:

- prefer the latest version >=x and <y
- accept anything >=x and <y, must not use anything <x or >= y

One option is to simply provide factory methods for these 2 constraints:

```
dependencies {
   implementation group: 'abc', module: 'somelib', version: requireVersion('x')
   implementation group: 'abc', module: 'somelib', version: requireVersionIn('x', 'y')
   implementation ('abc:someLib') {
        version = requireVersion('x')
   }
}
```

These constraints are really just composites that compose up some other simpler constraints. We have some other use cases coming up that require these simpler constraints:

- "I prefer x" and "I prefer from x up to y", used in addition to some other constraints.
- "must not use x" and "must not use x up to y", used in addition to some other constraints.

So another approach is to provide factory methods for the simple constraints and ways to compose them. The constraints we have are:

- accept(p) where p is a predicate that matches component metadata (not a "constraint"). Any component instance that matches the predicate is accepted. Does not express an opinion about preference.
- prefer(p) where p is a predicate that matches component metadata. Any component instance that matches the predicate is preferred _and_ implicitly accepted.
- both(a, b) where a and b are constraints (not "predicates"), which accepts the intersection of both a and b and prefers the intersection of both a and b.

The predicates we have are something like:

- 'x' which matches component with version x only.
- various ranges, eg '[x,y]' which matches any component with version in the given range.
- latest(p) which matches the latest version of components that match predicate p.
- not(p).
- anyWithStatus('x'), which matches any component with status >= 'x'
- anyWithVersionStartingWith('x'), which I'd be happy to see disappear.
- later, anyWithAttribute(x, y), which matches any component whose x attribute is compatible with value y.

Everything else is just combinations of this:

- `version='x'` == both(accept('x'), prefer('x'))
- `version='[x,y]` == both(accept('[x,]'), prefer(latest('[x,y]'))
- "I require x" == prefer('x')
- "I prefer from x up to y and I don't work with z" == both(prefer('[x,y]), accept(not('z'))

And so on.

**Comment by oehme** [comment]
Sounds like a good direction to go. How about we start with `require(p)` and later add the other methods like `prefer()` and `both()` (or `and()`).

**Comment by bigdaz** [comment]
@melix We will need some release notes and documentation for this feature

**Comment by melix** [comment]
Shouldn't we wait until we have an end-to-end story? From DSL to resolving?

**Comment by bigdaz** [comment]
Yeah I guess we haven't really discussed how we plan to 'release' this functionality. I suppose developing in 'stealth mode' might be fine for now.

Would be be fine with `VersionConstraint` being part of our DSL but not announcing it for 4.4?

**Comment by melix** [comment]
Yup.

**Comment by melix** [comment]
Closing as delivered in #3305
